include/SDL3_ttf/SDL_ttf.h

```diff
@@ -1190,6 +1222,195 @@ extern SDL_DECLSPEC SDL_Surface * SDLCALL TTF_RenderText_LCD_Wrapped(TTF_Font *f
+ *
+ * \returns a TTF_TextEngine object or NULL on failure; call SDL_GetError() for more information.
+ *
+ * \threadsafety It is safe to call this function from any thread.
+ *
+ * \since This function is available since SDL_ttf 3.0.0.
+ *
+ * \sa TTF_DestroySurfaceTextEngine
+ * \sa TTF_DrawSurfaceText
+ */
+extern SDL_DECLSPEC TTF_TextEngine * SDLCALL TTF_CreateSurfaceTextEngine(void);
+
+/**
+ * Draw text to an SDL surface.
+ *
+ * \param text the text to draw.
+ * \param x the x coordinate in pixels, positive from the left edge towards the right.
+ * \param y the y coordinate in pixels, positive from the top edge towards the bottom.
+ * \param surface the surface to draw on.
+ * \returns true on success or false on failure; call SDL_GetError() for more
+ *          information.
+ *
+ * \threadsafety This function should be called on the thread that created the text.
+ *
+ * \since This function is available since SDL_ttf 3.0.0.
+ *
+ * \sa TTF_CreateSurfaceTextEngine
+ * \sa TTF_CreateText
+ * \sa TTF_CreateText_Wrapped
+ */
+extern SDL_DECLSPEC bool SDLCALL TTF_DrawSurfaceText(TTF_Text *text, int x, int y, SDL_Surface *surface);
+
+/**
+ * Destroy a text engine created for drawing text on SDL surfaces.
+ *
+ * All text created by this engine should be destroyed before calling this function.
+ *
+ * \param engine a TTF_TextEngine object created with TTF_CreateSurfaceTextEngine().
+ *
+ * \threadsafety This function should be called on the thread that created the engine.
+ *
+ * \since This function is available since SDL_ttf 3.0.0.
+ *
+ * \sa TTF_CreateSurfaceTextEngine
+ */
+extern SDL_DECLSPEC void SDLCALL TTF_DestroySurfaceTextEngine(TTF_TextEngine *engine);
+
+/**
+ * Create a text engine for drawing text on an SDL renderer.
+ *
+ * \param renderer the renderer to use for creating textures and drawing text.
+ * \returns a TTF_TextEngine object or NULL on failure; call SDL_GetError() for more information.
+ *
+ * \threadsafety This function should be called on the thread that created the renderer.
+ *
+ * \since This function is available since SDL_ttf 3.0.0.
+ *
+ * \sa TTF_DestroyRendererTextEngine
+ */
+extern SDL_DECLSPEC TTF_TextEngine * SDLCALL TTF_CreateRendererTextEngine(SDL_Renderer *renderer);
+
+/**
+ * Draw text to an SDL renderer.
+ *
+ * \param text the text to draw.
+ * \param x the x coordinate in pixels, positive from the left edge towards the right.
+ * \param y the y coordinate in pixels, positive from the top edge towards the bottom.
+ * \returns true on success or false on failure; call SDL_GetError() for more
+ *          information.
+ *
+ * \threadsafety This function should be called on the thread that created the text.
+ *
+ * \since This function is available since SDL_ttf 3.0.0.
+ *
+ * \sa TTF_CreateRendererTextEngine
+ * \sa TTF_CreateText
+ * \sa TTF_CreateText_Wrapped
+ */
+extern SDL_DECLSPEC bool SDLCALL TTF_DrawRendererText(TTF_Text *text, float x, float y);
```

Comment:
It's a little weird you can get a TTF_Text from TTF_CreateText that won't work with this, if it's from a surface-based (or some other future thing) engine. At a minimum, the docs should say this, but maybe the interface needs to change. If we don't think there will be other engines in the future, we should probably have a TTF_RendererText type so you can't mix between engines.

(But even there, we need to be document "this can't be used with a _different_ SDL_Renderer, even if this TTF_Text came from a renderer-based engine," because the underlying SDL_Texture isn't shared.)

Reply:
I was thinking the function naming would make it obvious, but we can definitely be explicit that the TTF_DrawXXXText() functions only work with text returned from TTF_CreateText() given TTF_XXXTextEngine. This is going to come up again in the GPU text engine shortly, and who knows how people will extend this in the future.

Reply:
Maybe just a generic TTF_DrawText() and SDL_ttf knows what engine is supposed to be used and where the text is going? And then we don't have to add a new TTF_Draw*Text function for each new implementation?

Reply:
No, because the drawing parameters are different depending on the underlying use case. The surface one takes a destination surface parameter, the renderer one has an implicit renderer but uses float coordinates, the GPU one ... who knows?

I went ahead and documented this better though, thanks!